### ISSUE TITLE
[BUG] Changes to Dynamic Theme get reverted/deleted

### ISSUE THREAD

**Issue by Gitoffthelawn** [darkreader/darkreader#4468]
### **Describe the bug**

When I use DR's dev tools to modify the default Dynamic Theme, the changes get reverted.

### **Steps to reproduce**

- Set Firefox to delete all items (cache, etc.) on close
- Disable syncing DR dynamic theme updates
- Using DR dev tools (v5 preview), add CSS rulesets to the `*` section of the dynamic theme
- Close Firefox
- Reopen Firefox
- Inspect the DR dynamic theme

### **Expected behavior**

Changes to dynamic theme not lost

### **Actual behavior**

All changes to dynamic theme get reverted

### **System Information:**

- OS: Windows
- Browser: Firefox 83
- Dark Reader version: Version 5 Preview (4.9.26)

### **Additional context**

Although this looks like a bug, it could possibly be an unintended consequence of not conveying when (and why) the dynamic theme would get reverted.

The bottom line is that I would like to be able to make changes to the dynamic theme and not lose those changes.  I have not submitted these changes to be applied to all Dark Reader users because they may only be applicable to specific OS themes, and possibly only specific browsers.  The changes involve modifying the CSS for all `input`, `select`, and `textarea` HTML elements.  For example:

`input, select, textarea {
    color: var(--darkreader-selection-text) !important; }`

**Comment by Gusted** [comment]
Hi @Gitoffthelawn!

You mention that you set Firefox to delete all items on close, this could be possible the culprit, but I'm unsure if Firefox deletes the `storage.local` for extension. Could you check with another extension e.g. uBlock origin if you change settings they will stay after re-opening Firefox.

The CSS should stay applied even after a restart etc. into the devtools with an exception when a new update of dark reader is pushed.

Regards,
Gusted

**Comment by Gitoffthelawn** [comment]
Hi Gusted!

I gave it a try with uBlock Origin, and it maintains all its settings between Firefox sessions.  I tried about a dozen other extensions as well, and they all maintain their settings too.  I was not able to find any other extensions with this issue.

As a side note, it would be great for Dark Reader to be able to maintain changes to the dynamic theme even after DR extension updates and dynamic theme sync updates.  I'm thinking perhaps a "preface" section within the dynamic theme for user modifications that override default dynamic theme CSS.  That "preface" section would always be maintained, even during updates and sync.  What do you think?

**Comment by Gusted** [comment]
> I gave it a try with uBlock Origin, and it maintains all its settings between Firefox sessions. I tried about a dozen other extensions as well, and they all maintain their settings too. I was not able to find any other extensions with this issue.

Good to know, just to exclude another factor what if you try to change Dark Reader's settings e.g. grayscale or sepia.

> As a side note, it would be great for Dark Reader to be able to maintain changes to the dynamic theme even after DR extension updates and dynamic theme sync updates. I'm thinking perhaps a "preface" section within the dynamic theme for user modifications that override default dynamic theme CSS. That "preface" section would always be maintained, even during updates and sync. What do you think?

Their are some open issues about this #1477 #1856 and personally think it's a good, however it requires a lot of new logic inside dark reader to add support for CSS customization and is on the backslash for now.

**Comment by Gitoffthelawn** [comment]
> Good to know, just to exclude another factor what if you try to change Dark Reader's settings e.g. grayscale or sepia.

Are you referring to changing the grayscale or sepia of DR's dynamic theme?

**Comment by Gusted** [comment]
> Are you referring to changing the grayscale or sepia of DR's dynamic theme?

Yes

**Comment by Gitoffthelawn** [comment]
Great. I just wanted to be sure. ;)

Yes, the setting for sepia in the dynamic theme is maintained.

I think my bug report might possibly be a touch misleading.  I'm not sure whether or not the first step "Set Firefox to delete all items (cache, etc.) on close" is a red herring.  Why?  Because this issue does not happen every time.

I'm thinking it's possible this issue happens only when Dark Reader does something that is not visible to the user.  What do you think?  For example, IIRC, DR checks for notifications once and a while from a server.  Is it possible that the dynamic theme is reset at that time, or when some other *occasional* action is taken by DR?

**Comment by Gusted** [comment]
Hi @Gitoffthelawn!

As I remember correctly `localStorage` is also being removed when Firefox closes.
I just found out that devTools still uses localStorage 
![image](https://user-images.githubusercontent.com/25481501/101761740-6c563980-3ad4-11eb-8a5a-4d057a2af6bf.png)
to store it's data, so most likely when closing Firefox this data is also removed and dark reader will use the default one.

Regards,
Gusted

**Comment by Gusted** [comment]
> As I remember correctly `localStorage` is also being removed when Firefox closes.

Actually not after testing. So their seems to be some resetting? Do you have any error when:
- Going to about:debugging#/runtime/this-firefox
- Click on inspect 
![image](https://user-images.githubusercontent.com/25481501/101762499-6ca30480-3ad5-11eb-9ce5-b06f5f8fadb4.png)
- Goes to console 
![image](https://user-images.githubusercontent.com/25481501/101762534-7593d600-3ad5-11eb-9538-bdcb29abb7f6.png)

**Comment by Gusted** [comment]
I just tested on a profile with the same steps, but doesn't seems to reproduce.

Do you get an error when
- Go to about:debugging#/runtime/this-firefox 
- Click on inspect by Dark Reader
- Goes to console 
![image](https://user-images.githubusercontent.com/25481501/101762681-a247ed80-3ad5-11eb-8b0b-e9e2ed979d18.png)

**Comment by Gitoffthelawn** [comment]
Eek!  Local storage!

Yes, I see a bunch of errors in the console.  Let me create a screenshot for you...

**Comment by Gitoffthelawn** [comment]
Okay, I did some more investigating.  I think this particular issue is *not* a bug in Dark Reader.  I was unaware that it stores any data in Local Storage, and I do manually delete `webappsstore.sqlite` on occasion.   The last time I checked, I didn't have anything still using Local Storage, but I didn't check after I started using Dark Reader again.  Definitely time for Dark Reader to move away from Local Storage!  Should be simple, no?

That said, I did find some console errors that we maybe we can figure out together.  The only ones relevant (or possibly relevant) to Dark Reader are:

```
xx:xx:11.678 TypeError: this.contentWindow is null		ExtensionCommon.jsm:499:15
    openConduit resource://gre/modules/ExtensionCommon.jsm:499
    ChildAPIManager resource://gre/modules/ExtensionChild.jsm:745
    <anonymous> resource://gre/modules/ExtensionContent.jsm:961
    get resource://gre/modules/ExtensionCommon.jsm:157
    ContentScriptContextChild resource://gre/modules/ExtensionContent.jsm:867
    get resource://gre/modules/ExtensionCommon.jsm:157
    ContentScriptContextChild resource://gre/modules/ExtensionContent.jsm:872
    get resource://gre/modules/Schemas.jsm:192
    <anonymous> moz-extension://.../inject/index.js:1428
    <anonymous> moz-extension://.../inject/index.js:3800
    inject resource://gre/modules/ExtensionContent.jsm:568
    AsyncFunctionNext self-hosted:674
xx:xx:11.683 uncaught exception: Object				2
xx:xx:13.328
  /''''\
 (0)==(0)
/__||||__\
Welcome to Dark Reader!						index.js:3136:13
xx:xx:13.875 Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/		iconupdater.js:47:8
xx:xx:14.156 loaded 
Object { enabled: true, theme: {…}, presets: (3) […], customThemes: (19) […], siteList: (85) […], siteListEnabled: [], applyToListedOnly: false, changeBrowserTheme: false, notifyOfNews: false, syncSettings: false, … }
			index.js:2899:21
xx:xx:14.603 Uncaught (in promise) Error: Missing host permission for the tab, and any iframes	undefined
xx:xx:51.272 Uncaught (in promise) Error: Missing host permission for the tab			undefined
```

- The `TypeError` seems reasonably likely to be a DR bug.  Probably worth investigating.
- The `uncaught exception: Object` (2 of them, consecutively) may be a DR error or not.  Hard for me tell with this report.
- The `Synchronous XMLHttpRequest` error...  What do you think?
- The `Uncaught (in promise)` errors.  No idea at this point if it has anything to do with DR.  Not much to go on right now.  Any ideas?
- I also noticed that DR shows `notifyOfNews: false`, but according to logs, it attempts to connect to the DR news server regularly.  I'm thinking that's probably a bug.  You?

**Comment by Gusted** [comment]
> Okay, I did some more investigating. I think this particular issue is _not_ a bug in Dark Reader. I was unaware that it stores any data in Local Storage, and I do manually delete `webappsstore.sqlite` on occasion. The last time I checked, I didn't have anything still using Local Storage, but I didn't check after I started using Dark Reader again. Definitely time for Dark Reader to move away from Local Storage! Should be simple, no?

Well, it's original idea of the devtools was to temporal check if fixes that were made is working without any idea to have it as an permanent option. I rather to have that idea and work on a separate place for permanent CSS fixes that doesn't depend on localStorage.

>     * The `TypeError` seems reasonably likely to be a DR bug.  Probably worth investigating.

It's being caused in a native code from firefox. https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ExtensionCommon.jsm#499

>     * The `uncaught exception: Object` (2 of them, consecutively) may be a DR error or not.  Hard for me tell with this report.

I've seen them before but never got the exact reason why.



>     * The `Synchronous XMLHttpRequest` error...  What do you think?

Being caused in https://github.com/M-Reimer/togglefonts/blob/master/iconupdater.js#L47


>     * The `Uncaught (in promise)` errors.  No idea at this point if it has anything to do with DR.  Not much to go on right now.  Any ideas?

I have 0 idea, might be another extension that is causing this. Dark reader is by default requestion to access all URL's.



>     * I also noticed that DR shows `notifyOfNews: false`, but according to logs, it attempts to connect to the DR news server regularly.  I'm thinking that's probably a bug.  You?

This is indeed a problem and looked in the code that their is some trickery going around here, I will have a look why.

Regards,
Gusted

**Comment by Gitoffthelawn** [comment]
First, a quick apology. I only have a moment to respond, so please forgive my brevity.

> Well, it's original idea of the devtools was to temporal check if fixes that were made is working without any idea to have it as an permanent option. I rather to have that idea and work on a separate place for permanent CSS fixes that doesn't depend on localStorage.

Sounds good.  Would love to see that developed soon.  In the meantime, would it be simple to move devtools out o localStorage?

> It's being caused in a native code from firefox. https://searchfox.org/mozilla-central/source/toolkit/components/extensions/ExtensionCommon.jsm#499

Interesting.  I didn't look at that one long enough.  You're right, of course.  Do you think it's an actual Firefox bug?  Would you link to report it?

> I've seen them before but never got the exact reason why.

Same here.

> Being caused in https://github.com/M-Reimer/togglefonts/blob/master/iconupdater.js#L47

Is that something Dark Reader is currently calling (I don't use that extension)?  Has the bug been reported upstream?

> I have 0 idea, might be another extension that is causing this. Dark reader is by default requestion to access all URL's.

It's on my list of stuff to investigate.   No idea at this point whether it is related to Dark Reader.

> This is indeed a problem and looked in the code that their is some trickery going around here, I will have a look why.

Great. Thanks.

**Comment by Gusted** [comment]
> First, a quick apology. I only have a moment to respond, so please forgive my brevity.

No problem!

> Sounds good. Would love to see that developed soon. In the meantime, would it be simple to move devtools out o localStorage?

If I will be touching the storage of Devtools in the meanwhile before it's reviewed and merged and shipped into the next version I probably have a working prototype


> Interesting. I didn't look at that one long enough. You're right, of course. Do you think it's an actual Firefox bug? Would you link to report it?

I'm currently not bother to report those bugs as I don't know the exact cause of it.



> Is that something Dark Reader is currently calling (I don't use that extension)? Has the bug been reported upstream?

I've never seen that extension before, neither does Dark Reader call it. It's not a bug rather a warning that `synchronous` XMLHttpRequest is not supported any more/deprecated. 

> It's on my list of stuff to investigate. No idea at this point whether it is related to Dark Reader.

Seeing somehow having other extension on our context script, I have a slight suspicion.

Expect in the next few days a PR to fix this problem.

**Comment by Gusted** [comment]
>     * I also noticed that DR shows `notifyOfNews: false`, but according to logs, it attempts to connect to the DR news server regularly.  I'm thinking that's probably a bug.  You?

#4472

**Comment by AG-7219** [comment]
@Gitoffthelawn hey, were you able to fix the issue, I have been using firefox and the dynamic themes are reverted back as soon as I close the browser. I also tried shifting to static themes but it switches to dynamic themes automatically when I open the browser after closing. it seems like DR doesn't store any information about the previous session. Besides I get these errors during debugging
![image](https://user-images.githubusercontent.com/76441660/141637411-2c785119-caa2-4dfa-9255-e6a92de68258.png)
Besides my firefox doesnt delete the site data
![image](https://user-images.githubusercontent.com/76441660/141638374-c14b5cbe-f2ea-4be0-baef-8487e880c670.png)

OS : Ubuntu 20.04
browser : firefox developer edition Version 95.0b6 (64-bit)

**Comment by Mds92** [comment]
I have the same problem in Windows 10 / Chrome Version 95.0.4638.69 (Official Build) (64-bit)
Any solutions?

**Comment by Gitoffthelawn** [comment]
> 
> 
> @Gitoffthelawn hey, were you able to fix the issue, I have been using firefox and the dynamic themes are reverted back as soon as I close the browser...

I thought the issue was resolved, but I may not be remembering correctly because, for the most part, I stopped using Dark Reader's editor.  I found it much faster to add style filters to uBO to fix dynamic dark mode issues instead of using Dark Reader's dynamic theme editor.

For changes to the default dynamic mode that I thought would benefit others, I started creating pull requests in this repo so everyone can benefit.  I'm still doing that once in a while, but I've slowed down because those pull requests must support dark *and* light dynamic modes, and I only use Dark Reader for dark modes.  I have hundreds of dynamic dark mode overrides, and it was taking too much time to support dynamic light mode too (do more than 2 people use that?).

**Comment by Gusted** [comment]
See #7359

**Comment by AG-7219** [comment]
I see that the PR's are merged, when can we expect a release for the same?

**Comment by Gusted** [comment]
Mostly releases are at the beginning of the month.

**Comment by AG-7219** [comment]
Okay, thanks a lot, hope to see it fixed soon.

**Comment by ars4l4n** [comment]
> Okay, thanks a lot, hope to see it fixed soon.

me too

**Comment by ipcjs** [comment]
When will the new version be released?

**Comment by trebory6** [comment]
Also wondering about this. I'm on Chrome and it's doing the same exact thing. Incredibly annoying.

Our company uses tools on an internal intranet, and I've been using this to create a darkmode theme for them that our whole team uses.  We can't just copy and paste our custom CSS into the Dynamic Theme Editor every time we restart the browser. 

It's so niche we can't add to the public list either.

**Comment by Gitoffthelawn** [comment]
@ipcjs @trebory6 Is everything working correctly for you now?

**Comment by ipcjs** [comment]
Yes

**Comment by AG-7219** [comment]
the editor does remember the css but i still have to open it and click on apply everytime i restart the browser for the changes to take place. was this intended or some issue with my browser?

**Comment by Gusted** [comment]
> the editor does remember the css but i still have to open it and click on apply everytime i restart the browser for the changes to take place. was this intended or some issue with my browser?

That was another issue, it's fixed in the next release.

**Comment by Gitoffthelawn** [comment]
As I jokingly say to a good friend of mine, "Your issues have issues." :)

**Comment by ars4l4n** [comment]
even after your latest update that fixed the issue of advanced settings being reverted after a browser restart, the settings got reset at some point again. I didn't pay attention to what caused it though. I can just say that I'm sure I didn't reinstall my browser or anything alike.
running the latest Chrome update, Version 97.0.4692.99 on Windows 10 64bit Version 1903 and DarkReaders' December update.

**Comment by Gusted** [comment]
> even after your latest update that fixed the issue of advanced settings being reverted after a browser restart, the settings got reset at some point again. I didn't pay attention to what caused it though. I can just say that I'm sure I didn't reinstall my browser or anything alike.
> Running the latest Chrome update, Version 97.0.4692.99 on Windows 10 64bit Version 1903 and DarkReaders' December update
> Why does this happen? Do you perhaps know a sort of meta-solution that can prevent this? Like running some forked versions of extensions locally or something like that?

I'm not able to debug a problem that happens out of thin air.

This issue has been fixed for a while so I'm closing this issue as well now. If anyone feels like they see this problem, open up a new issue about it.
